clients/src/main/java/org/apache/kafka/common/requests/TxnOffsetCommitRequest.java

```diff
@@ -78,13 +88,36 @@ public Builder(final String transactionalId,
                             .setMemberId(memberId)
                             .setGenerationId(generationId)
                             .setGroupInstanceId(groupInstanceId.orElse(null));
+            this.autoDowngrade = autoDowngrade;
         }
 
         @Override
         public TxnOffsetCommitRequest build(short version) {
+            if (version < 3 && groupMetadataSet()) {
+                if (autoDowngrade) {
+                    log.trace("Downgrade the request by resetting group metadata fields: " +
+                                  "[member.id:{}, generation.id:{}, group.instance.id:{}], because broker " +
+                                  "only supports TxnOffsetCommit version {}. Need " +
+                                  "v3 or newer to enable this feature",
+                        data.memberId(), data.generationId(), data.groupInstanceId(), version);
+
+                    data.setGenerationId(JoinGroupRequest.UNKNOWN_GENERATION_ID)
+                        .setMemberId(JoinGroupRequest.UNKNOWN_MEMBER_ID)
+                        .setGroupInstanceId(null);
+                } else {
+                    throw new UnsupportedVersionException("Broker unexpectedly " +
+                        "doesn't support group metadata commit API on version " + version);
+                }
+            }
             return new TxnOffsetCommitRequest(data, version);
         }
 
+        private boolean groupMetadataSet() {
+            return !data.memberId().equals(JoinGroupRequest.UNKNOWN_MEMBER_ID) ||
+                       data.generationId() != JoinGroupRequest.UNKNOWN_GENERATION_ID ||
+                       data.groupInstanceId() != null;
```

Comment:
groupInstanceId field should never be null, I think we should check against `Optional.empty()`.